### PR TITLE
Replace error-chain for client error

### PR DESCRIPTION
**Pull request by ascjones**
Starts to address #1547, just for the client `Error` to start with. It replaces the error_chain macro with:

- Manually implemented `Error` (including `source`)
- Derived `Display` from [`derive-more`](https://docs.rs/derive_more/0.14.0/derive_more/) to handle display boilerplate.
- Derive `From` impls: however [only works](https://jeltef.github.io/derive_more/derive_more/from.html#enums) for variants with unique types. So for example there are multiple variants with `String`, in this case we must manually define the `From` impl.

If we think this is a good approach  we can apply it to the remaining errors in follow up PRs.

### CODE REVIEW

File: core/client/src/error.rs [hunk 1]
```diff
@@ -16,141 +16,119 @@
...
+}
 
-		/// Last finalized block not parent of current.
-		NonSequentialFinalization(s: String) {
-			description("Did not finalize blocks in sequential order."),
-			display("Did not finalize blocks in sequential order."),
-		}
+impl From<String> for Error {
+	fn from(s: String) -> Self {
+		Error::Msg(s)
```
[the remark this answers is not in the document]
Comment (ascjones) [reply core/client/src/error.rs]: I've just added `From` derive, however that only works for variants where the type is unique - which is not true for Strings in this enum.

I can add a macro for this in a follow up PR converting the other errors.

File: core/client/src/error.rs [hunk 2]
```diff
@@ -16,159 +16,116 @@
...
+}
 
-		/// Hash that is required for building CHT is missing.
-		MissingHashRequiredForCHT(cht_num: u64, block_number: u64) {
-			description("missed hash required for building CHT"),
-			display("Failed to get hash of block#{} for building CHT#{}", block_number, cht_num),
+impl error::Error for Error {
+	fn source(&self) -> Option<&(error::Error + 'static)> {
+		match self {
+			Error::Consensus(e) => Some(e),
```
Comment (bkchr) [inline core/client/src/error.rs]: Can we not add `Blockchain` and `Execution` here as well?

File: core/client/src/error.rs [hunk 3]
```diff
@@ -16,159 +16,116 @@
...
+	fn from(s: &'a str) -> Self {
+		Error::Msg(s.into())
 	}
 }
 
 impl Error {
 	/// Chain a blockchain error.
 	pub fn from_blockchain(e: Box<std::error::Error + Send>) -> Self {
-		ErrorKind::Blockchain(e).into()
+		Error::Blockchain(e).into()
```
Comment (bkchr) [inline core/client/src/error.rs]:
```suggestion
		Error::Blockchain(e)
```

File: core/client/src/error.rs [hunk 4]
```diff
@@ -16,159 +16,116 @@
...
 	}
 
 	/// Chain a state error.
 	pub fn from_state(e: Box<state_machine::Error + Send>) -> Self {
-		ErrorKind::Execution(e).into()
+		Error::Execution(e).into()
```
Comment (bkchr) [inline core/client/src/error.rs]:
```suggestion
		Error::Execution(e)
```

File: core/client/src/in_mem.rs [hunk 5]
```diff
@@ -207,7 +207,7 @@ impl<Block: BlockT> Blockchain<Block> {
 	pub fn set_head(&self, id: BlockId<Block>) -> error::Result<()> {
 		let header = match self.header(id)? {
 			Some(h) => h,
-			None => return Err(error::ErrorKind::UnknownBlock(format!("{}", id)).into()),
+			None => return Err(error::Error::UnknownBlock(format!("{}", id)).into()),
```
Comment (bkchr) [inline core/client/src/in_mem.rs]:
```suggestion
			None => return Err(error::Error::UnknownBlock(format!("{}", id))),
```

File: core/client/src/in_mem.rs [hunk 6]
```diff
@@ -258,7 +258,7 @@ impl<Block: BlockT> Blockchain<Block> {
 	fn finalize_header(&self, id: BlockId<Block>, justification: Option<Justification>) -> error::Result<()> {
 		let hash = match self.header(id)? {
 			Some(h) => h.hash(),
-			None => return Err(error::ErrorKind::UnknownBlock(format!("{}", id)).into()),
+			None => return Err(error::Error::UnknownBlock(format!("{}", id)).into()),
```
Comment (bkchr) [inline core/client/src/in_mem.rs]:
```suggestion
			None => return Err(error::Error::UnknownBlock(format!("{}", id))),
```

File: core/client/src/in_mem.rs [hunk 7]
```diff
@@ -416,12 +416,12 @@ impl<Block: BlockT> light::blockchain::Storage<Block> for Blockchain<Block>
 
 	fn header_cht_root(&self, _cht_size: u64, block: NumberFor<Block>) -> error::Result<Block::Hash> {
 		self.storage.read().header_cht_roots.get(&block).cloned()
-			.ok_or_else(|| error::ErrorKind::Backend(format!("Header CHT for block {} not exists", block)).into())
+			.ok_or_else(|| error::Error::Backend(format!("Header CHT for block {} not exists", block)).into())
```
Comment (bkchr) [inline core/client/src/in_mem.rs]:
```suggestion
			.ok_or_else(|| error::Error::Backend(format!("Header CHT for block {} not exists", block)))
```

File: core/client/src/in_mem.rs [hunk 8]
```diff
@@ -416,12 +416,12 @@ impl<Block: BlockT> light::blockchain::Storage<Block> for Blockchain<Block>
...
 	}
 
 	fn changes_trie_cht_root(&self, _cht_size: u64, block: NumberFor<Block>) -> error::Result<Block::Hash> {
 		self.storage.read().changes_trie_cht_roots.get(&block).cloned()
-			.ok_or_else(|| error::ErrorKind::Backend(format!("Changes trie CHT for block {} not exists", block)).into())
+			.ok_or_else(|| error::Error::Backend(format!("Changes trie CHT for block {} not exists", block)).into())
```
Comment (bkchr) [inline core/client/src/in_mem.rs]: Okay, there are a lot more `into()` calls that now can be dropped.

File: core/client/src/error.rs (hunk 2 above)
Comment (bkchr) [reply core/client/src/error.rs]: @ascjones ?

Comment (ascjones) [reply core/client/src/error.rs]: I've allowed `Blockchain` by changing it to the concrete type (it's recursive).

Not sure about `Execution` error, it's a custom trait and not `std::error::Error`. https://github.com/paritytech/substrate/blob/3a4901a78f627d689673597e5b8fca761d954ae3/core/state-machine/src/lib.rs#L61

**Pull request merged**